### PR TITLE
Always set the generated flag to false when an avatar is being set

### LINKED ISSUE #18481: Avatar endpoint returns wrong response code
According to https://github.com/nextcloud/server/blame/12bdb95c1a40170652570decef7f466e4b97aa65/core/Controller/AvatarController.php#L145, the server should return 200 if a user has an uploaded avatar (rather than generated one). 

This doesn't seem to be the case as the server returns 201 regardless of the type of avatar user has.

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #18481

### CODE REVIEW

File: lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	private $connection;
+
+	public function __construct(IConfig $config,
+								IDBConnection $connection) {
+		$this->config = $config;
+		$this->connection = $connection;
+	}
+
+	public function getName(): string {
+		return 'Clear access cache of projects';
```
Comment (kesselb) [inline lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php]: :thinking:

File: lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	private function shouldRun(): bool {
+		$versionFromBeforeUpdate = $this->config->getSystemValue('version', '0.0.0.0');
+		return version_compare($versionFromBeforeUpdate, '18.0.0.5', '<=');
+	}
+
+	public function run(IOutput $output): void {
+		if ($this->shouldRun()) {
+			$query = $this->connection->getQueryBuilder();
+			$query->delete('preferences')
```
Comment (kesselb) [inline lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php]: https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/Avatar/UserAvatar.php#L294-L297

Scares me a bit but will probably work due to above code.

**Review by kesselb — APPROVED** [review]

File: lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php (hunk 2 above)
Comment (nickvergessen) [reply lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php]: Yes, I checked this before and it worked, even when accessing an avatar directly as a first thing.

File: lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php]: :laughing:

File: lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php (hunk 2 above)
Comment (kesselb) [reply lib/private/Repair/NC18/ResetGeneratedAvatarFlag.php]: OK. A bit confusing to have generated and generated but the app value should be faster in most cases. Cache for a state ;)

**Comment by kesselb** [comment]
I still wonder where the `generated` file is deleted. But this pr fixes another issue. If the flag generated is true the delete button for an avatar is missing.

**Comment by kesselb** [comment]
> I still wonder where the `generated` file is deleted.

https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/Avatar/UserAvatar.php#L104

Found it. It's probably safe to remove the other remove statement with the try-catch around. But there is no lock for the folder so another process could create a new generated avatar in the meantime. I don't know ;)

**Comment by nickvergessen** [comment]
Yeah I gues that is exactly the issue.

**Review by rullzer — APPROVED** [review]

**Pull request merged**